src/test/scala/examples/LoadMemoryFromFileSpec.scala

```diff
@@ -0,0 +1,77 @@
+// See LICENSE for license details.
+
+package examples
+
+// See LICENSE for license details.
+
+import java.nio.file.{Files, Paths}
+import java.nio.file.StandardCopyOption.REPLACE_EXISTING
+
+import chisel3._
+import chisel3.iotesters.PeekPokeTester
+import chisel3.util.loadMemoryFromFile
+import firrtl.FileUtils
+import org.scalatest.{FreeSpec, Matchers}
+
+//noinspection TypeAnnotation
+class UsesMem(memoryDepth: Int, memoryType: Bits) extends Module {
+  val io = IO(new Bundle {
+    val address = Input(UInt(memoryType.getWidth.W))
+    val value   = Output(memoryType)
+    val value2  = Output(memoryType)
+  })
+
+  val memory = Mem(memoryDepth, memoryType)
+
+  loadMemoryFromFile(memory, "test_run_dir/load_mem_test/mem1.txt")
+
+  io.value := memory(io.address)
+
+  val low = Module(new UsesMemLow(memoryDepth, memoryType))
+
+  low.io.address := io.address
+  io.value2 := low.io.value
+}
+
+class UsesMemLow(memoryDepth: Int, memoryType: Data) extends Module {
+  val io = IO(new Bundle {
+    val address = Input(UInt(memoryType.getWidth.W))
+    val value   = Output(memoryType)
+  })
+
+  val memory = Mem(memoryDepth, memoryType)
+
+  loadMemoryFromFile(memory, "test_run_dir/load_mem_test/mem2.txt")
+
+  io.value := memory(io.address)
+}
+
+class LoadMemoryFromFileTester(c: UsesMem) extends PeekPokeTester(c) {
+  for(addr <- 0 until 8) {
+    poke(c.io.address, addr)
+    step(1)
+    println(f"peek from $addr ${peek(c.io.value)}%x ${peek(c.io.value2)}%x")
+    expect(c.io.value, addr)
+    expect(c.io.value2, 7 - addr)
+  }
+}
+
+class LoadMemoryFromFileSpec extends FreeSpec with Matchers {
+  "Users can specify a source file to load memory from" in {
+
+    val targetDirName = "test_run_dir/load_mem_test"
+    FileUtils.makeDirectory(targetDirName)
+
+    val path1 = Paths.get(targetDirName + "/mem1.txt")
+    val path2 = Paths.get(targetDirName + "/mem2.txt")
+    Files.copy(getClass.getResourceAsStream("/mem1.txt"), path1, REPLACE_EXISTING)
+    Files.copy(getClass.getResourceAsStream("/mem2.txt"), path2, REPLACE_EXISTING)
+
+    iotesters.Driver.execute(
+      args = Array("--backend-name", "verilator", "--target-dir", targetDirName, "--top-name", "load_mem_test"),
```

Comment:
I'm not sure of that, I think multiple backends are a more likely target for testers2. I don't think it's worth adding them here

Reply:
Sounds good to me. We'd probably have to think about how this would work. I think for your push on the backend plugin API, it would be interesting to think how memory loading could be abstracted for arbitrary backends. Definition itely future work.